File: test/test_api_comments.rb
require_relative "../lib/rgfa.rb"
require "test/unit"
TestAPI ||= Module.new

class TestAPI::Comments < Test::Unit::TestCase

  def test_initialize
   l = RGFA::Line::Comment.new(["hallo"])
   assert_equal("# hallo", l.to_s)
   l = RGFA::Line::Comment.new(["hallo", "\t"])
   assert_equal("#\thallo", l.to_s)
  end

  def test_fields
   l = RGFA::Line::Comment.new(["hallo"])
   assert_equal("hallo", l.content)
   assert_equal(" ", l.spacer)
   l.content = "hello"
   assert_equal("hello", l.content)
   assert_equal("# hello", l.to_s)
   l.spacer = "  "
   assert_equal("hello", l.content)
   assert_equal("#  hello", l.to_s)
  end

  def test_validation
   assert_raises(RGFA::FormatError) {RGFA::Line::Comment.new(["hallo\nhallo"])}
   assert_raises(RGFA::FormatError) {RGFA::Line::Comment.new(["hallo", "\n"])}
   assert_nothing_raised {
       RGFA::Line::Comment.new(["hallo", "\n"], validate: 0) }
   l = RGFA::Line::Comment.new(["hallo"])
   assert_nothing_raised {l.content = "hallo\n"}
   assert_raises(RGFA::FormatError) { l.to_s }
   l.content = "hallo"
   assert_nothing_raised {l.to_s}
   assert_nothing_raised {l.spacer = "\n"}
   assert_raises(RGFA::FormatError) { l.to_s }
   l = RGFA::Line::Comment.new(["hallo"], validate: 5)
   assert_raises(RGFA::FormatError) { l.content = "hallo\n" }
   assert_raises(RGFA::FormatError) { l.spacer = "\n" }
  end

  def test_from_string
    str = "# this is a comment"
    l = str.to_rgfa_line
    assert_equal(RGFA::Line::Comment, l.class)
    assert_equal(str[2..-1], l.content)
    assert_equal(" ", l.spacer)
    str = "#this is another comment"
    l = str.to_rgfa_line
    assert_equal(RGFA::Line::Comment, l.class)
    assert_equal(str[1..-1], l.content)
    assert_equal("", l.spacer)
    str = "#\t and this too"
    l = str.to_rgfa_line
    assert_equal(RGFA::Line::Comment, l.class)
    assert_equal(str[3..-1], l.content)
    assert_equal(str[1..2], l.spacer)
    str = "#: and this too"
    l = str.to_rgfa_line
    assert_equal(RGFA::Line::Comment, l.class)
    assert_equal(str[1..-1], l.content)
    assert_equal("", l.spacer)
  end

  def test_to_s
    str = "# this is a comment"
    l = str.to_rgfa_line
    assert_equal(str, l.to_s)
    str = "#this is another\tcomment"
    l = str.to_rgfa_line
    assert_equal(str, l.to_s)
    str = "#this is another\tcomment"
    l = str.to_rgfa_line
    l.spacer = " "
    assert_equal("# "+str[1..-1], l.to_s)
  end

  def test_tags
    assert_raises(RGFA::ValueError) {
      RGFA::Line::Comment.new(["hallo", " ", "zz:Z:hallo"]) }
    l = "# hallo zz:Z:hallo".to_rgfa_line
    assert_equal("hallo zz:Z:hallo", l.content)
    assert_raises(NoMethodError) { l.zz  }
    assert_raises(NoMethodError) { l.zz = 1 }
    assert_raises(RGFA::RuntimeError) { l.set(:zz, 1) }
    assert_nil(l.get(:zz))
  end

  def test_to_gfa1
    str = "# this is a comment"
    l = str.to_rgfa_line(version: :"2.0")
    assert_equal(RGFA::Line::Comment, l.class)
    assert_equal(:"2.0", l.version)
    assert_equal(str, l.to_s)
    assert_equal(:"2.0", l.to_gfa2.version)
    assert_equal(str, l.to_gfa2.to_s)
    assert_equal(:"1.0", l.to_gfa1.version)
    assert_equal(str, l.to_gfa1.to_s)
  end

  def test_to_gfa2
    str = "# this is a comment"
    l = str.to_rgfa_line(version: :"1.0")
    assert_equal(RGFA::Line::Comment, l.class)
    assert_equal(:"1.0", l.version)
    assert_equal(str, l.to_s)
    assert_equal(:"1.0", l.to_gfa1.version)
    assert_equal(str, l.to_gfa1.to_s)
    assert_equal(:"2.0", l.to_gfa2.version)
    assert_equal(str, l.to_gfa2.to_s)
  end

  def test_rgfa_comments
    gfa = RGFA.new
    c1 = "#this is a comment"
    c2 = "# this is also a comment"
    c3 = "#and \tthis too!"
    assert_nothing_raised { gfa << c1 }
    assert_nothing_raised { gfa << c2 }
    assert_nothing_raised { gfa << c3 }
    assert_equal([c1,c2,c3], gfa.comments.map(&:to_s))
    assert_equal(c1, gfa.comments[0].to_s)
    gfa.rm(gfa.comments[0])
    assert_equal([c2,c3], gfa.comments.map(&:to_s))
    gfa.comments[0].disconnect
    assert_equal([c3], gfa.comments.map(&:to_s))
  end

end

File: tests/test_api_comments.py
import unittest
import gfapy

class TestApiComments(unittest.TestCase):

  def test_initialize(self):
   l = gfapy.line.Comment("# hallo")
   self.assertEqual("# hallo", str(l))
   l = gfapy.line.Comment(["#", "hallo", "\t"])
   self.assertEqual("#\thallo", str(l))

  def test_fields(self):
   l = gfapy.line.Comment("# hallo")
   self.assertEqual("hallo", l.content)
   self.assertEqual(" ", l.spacer)
   l.content = "hello"
   self.assertEqual("hello", l.content)
   self.assertEqual("# hello", str(l))
   l.spacer = "  "
   self.assertEqual("hello", l.content)
   self.assertEqual("#  hello", str(l))

  def test_validation(self):
   with self.assertRaises(gfapy.FormatError):
     gfapy.line.Comment(["#", "hallo\nhallo"])
   with self.assertRaises(gfapy.FormatError):
     gfapy.line.Comment(["#", "hallo", "\n"])
   gfapy.line.Comment(["#", "hallo", "\n"], vlevel=0) # nothing raised
   l = gfapy.line.Comment(["#", "hallo"])
   l.content = "hallo\n" # nothing raised
   with self.assertRaises(gfapy.FormatError): str(l)
   l.content = "hallo"
   str(l) # nothing raised
   l.spacer = "\n" # nothing raised
   with self.assertRaises(gfapy.FormatError): str(l)
   l = gfapy.line.Comment(["#", "hallo"], vlevel=3)
   with self.assertRaises(gfapy.FormatError): l.content = "hallo\n"
   with self.assertRaises(gfapy.FormatError): l.spacer = "\n"

  def test_from_string(self):
    s = "# this is a comment"
    l = gfapy.Line(s)
    self.assertEqual(gfapy.line.Comment, l.__class__)
    self.assertEqual(s[2:], l.content)
    self.assertEqual(" ", l.spacer)
    s = "#this is another comment"
    l = gfapy.Line(s)
    self.assertEqual(gfapy.line.Comment, l.__class__)
    self.assertEqual(s[1:], l.content)
    self.assertEqual("", l.spacer)
    s = "#\t and this too"
    l = gfapy.Line(s)
    self.assertEqual(gfapy.line.Comment, l.__class__)
    self.assertEqual(s[3:], l.content)
    self.assertEqual(s[1:3], l.spacer)
    s = "#: and this too"
    l = gfapy.Line(s)
    self.assertEqual(gfapy.line.Comment, l.__class__)
    self.assertEqual(s[1:], l.content)
    self.assertEqual("", l.spacer)

  def test_to_s(self):
    s = "# this is a comment"
    l = gfapy.Line(s)
    self.assertEqual(s, str(l))
    s = "#this is another\tcomment"
    l = gfapy.Line(s)
    self.assertEqual(s, str(l))
    s = "#this is another\tcomment"
    l = gfapy.Line(s)
    l.spacer = " "
    self.assertEqual("# "+s[1:], str(l))

  def test_tags(self):
    with self.assertRaises(gfapy.ValueError):
      gfapy.line.Comment(["#", "hallo", " ", "zz:Z:hallo"])
    l = gfapy.Line("# hallo zz:Z:hallo")
    self.assertEqual("hallo zz:Z:hallo", l.content)
    self.assertEqual(None, l.zz)
    with self.assertRaises(gfapy.RuntimeError): l.zz = 1
    with self.assertRaises(gfapy.RuntimeError): l.set("zz", 1)
    self.assertEqual(None, l.get("zz"))

  def test_to_gfa1(self):
    s = "# this is a comment"
    l = gfapy.Line(s,version="gfa2")
    self.assertEqual(gfapy.line.Comment, l.__class__)
    self.assertEqual("gfa2", l.version)
    self.assertEqual(s, str(l))
    self.assertEqual("gfa2", l.to_gfa2().version)
    self.assertEqual(s, str(l.to_gfa2()))
    self.assertEqual("gfa1", l.to_gfa1().version)
    self.assertEqual(s, str(l.to_gfa1()))

  def test_to_gfa2(self):
    s = "# this is a comment"
    l = gfapy.Line(s,version="gfa1")
    self.assertEqual(gfapy.line.Comment, l.__class__)
    self.assertEqual("gfa1", l.version)
    self.assertEqual(s, str(l))
    self.assertEqual("gfa1", l.to_gfa1().version)
    self.assertEqual(s, str(l.to_gfa1()))
    self.assertEqual("gfa2", l.to_gfa2().version)
    self.assertEqual(s, str(l.to_gfa2()))

  def test_rgfa_comments(self):
    gfa = gfapy.Gfa()
    c1 = "#this is a comment"
    c2 = "# this is also a comment"
    c3 = "#and \tthis too!"
    gfa.add_line(c1) # nothing raised
    gfa.add_line(c2) # nothing raised
    gfa.add_line(c3) # nothing raised
    self.assertEqual([c1,c2,c3], [str(x) for x in gfa.comments])
    self.assertEqual(c1, str(gfa.comments[0]))
    gfa.rm(gfa.comments[0])
    self.assertEqual([c2,c3], [str(x) for x in gfa.comments])
    gfa.comments[0].disconnect()
    self.assertEqual([c3], [str(x) for x in gfa.comments])
